File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
from typing import List

class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        """
        Calculates the minimum number of extra characters needed to represent 
        the string `s` using words from the `dictionary`.

        Parameters:
        s (str): The input string to process.
        dictionary (List[str]): A list of words that can be used to represent the string.

        Returns:
        int: The minimum number of extra characters left after forming the string using the words from the dictionary.
        """
        sLength = len(s)
        # Initialize dp array where dp[i] represents minimum extra chars for s[:i]
        dp = [float('inf')] * (sLength + 1)
        dp[0] = 0  # No extra characters needed for an empty string

        # Convert the dictionary to a set for faster lookups
        wordSet = set(dictionary)

        # Iterate through the string to fill the dp array
        for i in range(1, sLength + 1):
            # By default, if no word matches, we have 1 extra character at dp[i]
            dp[i] = dp[i - 1] + 1
            # Check every possible start point of a word ending at index i
            for j in range(i):
                if s[j:i] in wordSet:
                    dp[i] = min(dp[i], dp[j])

        return dp[sLength]

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

/**
 * @brief Calculates the minimum number of extra characters needed to represent
 *        the string `s` using words from the `dictionary`.
 *
 * @param s (char*): The input string to process.
 * @param dictionary (char**): The dictionary containing words that can be used to represent the string.
 * @param dictionarySize (int): The size of the dictionary (number of words).
 * @return (int): The minimum number of extra characters left after forming the string using the words from the dictionary.
 */
int minExtraChar(char * s, char ** dictionary, int dictionarySize) {
    int sLength = strlen(s);
    int dp[sLength + 1]; // dp[i] stores the minimum extra chars for s[:i]
    
    // Initialize dp array
    dp[0] = 0;
    for (int i = 1; i <= sLength; i++) {
        dp[i] = dp[i - 1] + 1; // Default: no match means 1 extra character
        
        // Check all words in the dictionary for matching substring ending at i
        for (int j = 0; j < dictionarySize; j++) {
            int wordLength = strlen(dictionary[j]);
            if (i >= wordLength && strncmp(s + i - wordLength, dictionary[j], wordLength) == 0) {
                dp[i] = dp[i] < dp[i - wordLength] ? dp[i] : dp[i - wordLength];
            }
        }
    }
    
    return dp[sLength];
}